File: communicationAPI/src/main/java/eu/cybergeiger/communication/LocalApiFactory.java
package eu.cybergeiger.communication;

import ch.fhnw.geiger.localstorage.StorageException;
import java.util.HashMap;
import java.util.Map;

/**
 * <p>Implements a singleton pattern for local API.</p>
 */
public class LocalApiFactory {

  // TODO check for threadsafety as concurrenthashmap is not supported
  private static final Map<String, LocalApi> instances = new HashMap<>();

  /**
   * <p>Creates one instance only per id, but cannot guarantee it since LocalApi constructor cant be
   * private.</p>
   *
   * @param id          the id of the API to be retrieved
   * @param declaration the privacy declaration
   * @return the instance requested
   */
  public static LocalApi getLocalApi(String executor, String id, Declaration declaration)
      throws DeclarationMismatchException, StorageException {
    synchronized (instances) {
      if (!instances.containsKey(id)) {
        if (LocalApi.MASTER.equals(id)) {
          // create master
          instances.put(id, new LocalApi(executor, id, true, declaration));
        } else {
          instances.put(id, new LocalApi(executor, id, false, declaration));
        }
      }
    }
    LocalApi l = instances.get(id);
    if (declaration != null && l.getDeclaration() != declaration) {
      throw new DeclarationMismatchException();
    }
    return l;
  }

  /**
   * <p>Returns only existing and known instances.</p>
   *
   * @param id the id to be retrieved
   * @return the instance or null if not found
   */
  public static LocalApi getLocalApi(String id) {
    return instances.get(id);
  }
}

File: lib/src/communication/LocalApiFactory.dart
import 'dart:collection';

import 'Declaration.dart';
import 'DeclarationMismatchException.dart';
import 'LocalApi.dart';

/// Implements a singleton pattern for local API.
class LocalApiFactory {
  static final String MASTER_EXECUTOR = 'FIXME';

  static final Map<String, LocalApi> instances = HashMap();

  /// Creates or gets an instance.
  ///
  /// Will only create one [LocalApi] instance per plugin id without guarantee
  /// that this is the only instance since the [LocalApi] constructor cant be
  /// private.
  ///
  /// To stay compatible with the old Java API this is two methods in one.
  ///
  /// - Provide an executor ([executorOrId]), [id], and [declaration] is provided
  /// which creates a new instance in case none existing yet.
  /// - Provide only an id ([executorOrId]) is provided and it just returns
  /// the corresponding instance if it exists.
  ///
  /// Throws [DeclarationMismatchException] if the plugin has been registered previously and the
  /// declaration does not match and [StorageException] if registration failed.
  static LocalApi? getLocalApi(String executorOrId,
      [String? id, Declaration? declaration]) {
    if (id == null) {
      return instances[id];
    }
    // synchronized(instances, {
    if (!instances.containsKey(id)) {
      instances[id] =
          LocalApi(executorOrId, id, LocalApi.MASTER == id, declaration);
    }
    // });
    var l = instances[id]!;
    if ((declaration != null) && (l.getDeclaration() != declaration)) {
      throw DeclarationMismatchException();
    }
    return l;
  }
}
